### ISSUE TITLE
🦠 Model Request: Cyp3a4 Substrate

### ISSUE THREAD

**Issue by GemmaTuron** [ersilia-os/ersilia#422]
### Model Title

Cyp3a4 Substrate (TDC dataset)

### Publication

Hello @EstherIdabor!

As part of your Outreachy contribution, we have assigned you the dataset "cyp3a4 substrate" from the Therapeutics Data Commons to try and build a binary classification ML model. 
Please copy the provided Google Colab template and use this issue to provide updates on the progress. We'll value not only being able to build the model but also interpreting its results.

**Comment by EstherIdabor** [comment]
Thank you @GemmaTuron , on it

**Comment by EstherIdabor** [comment]
The steps I am taking/have taken in building a ML model to predict if a drug is a substrate of CYP3A4

1) Extensive research into the CYP3A4 enzyme: 
Pharmacokinetics(ADME); how the body interacts with administered drugs is defined in terms of site of administration, distribution to different part of the body, metabolism(how the drug is metabolized into harmless/inactive metabolites) and excretion.
In metabolism the focus is on how specialized enzymatic system breakdown drugs because this influences the duration of the drug action on the body as well as the intensity/concentration of the drug at a given time.
CYP3A4 is an enzyme under the CYP450 superfamily of enzymes found in the liver and intestine that oxidizes drugs or toxins so that they can be removed from the body.
Building a model that predict if a drug is metabolized/ a substrate of CYP3A4 is building a model that can be used to predict if the duration of drug action and and concentration is affected by the CYP3A4 enzyme.

2) Data Upload; 
In this step I installed the TDC(therapeutics Data Commons) package, then I imported ADME and got the specific assay assigned to me, I split the dataset using the `get_split( )` method in to train, validate and test which will be used  for training, validating and testing the model respectively.

**Comment by EstherIdabor** [comment]
3. **Data Analysis/Understanding my data**

After splitting my dataset into three datasets(train, validate, test) and ensuring they are saved to my drive, I then took a critical look at the dataset. by using the `len()` function, I got the value of the molecules in each dataset. In the train set they are a total number of **469 molecules**, in the validation set, they are a total number of **67 molecules**, in the test dataset, they are a total number of **134** molecules. 
I proceeded to get the number of actives in this case drugs that are a substrate of CYP3A4 and those that are not. 

- in the train set they are a total number of  **249** active molecules and **220** inactive molecules
- in the validation set, they are a total number of **39** active molecules and **28** inactive molecules
- in the test set, they are a total number of **67** active molecules and **67** inactive molecules

**Comment by GemmaTuron** [comment]
Good start @EstherIdabor !

**Comment by EstherIdabor** [comment]
> 

Thank you

**Comment by EstherIdabor** [comment]
> Good start @EstherIdabor !

Thank you

**Comment by EstherIdabor** [comment]
Using RDKit package I was able to visualize the molecular image of active(substrates) and inactive(non-substrate)

**Actives**
![download-rdkit-image](https://user-images.githubusercontent.com/70244748/198202144-8c96f327-e0b2-4b1f-bc33-7491d0761003.png)

**Inactives**
![rdkit-inactive-image](https://user-images.githubusercontent.com/70244748/198202192-aaa3b3bb-6c22-42f6-bb5a-31f68d1c4a08.png)

**Comment by EstherIdabor** [comment]
**Preparation for Modelling**

**Comment by GemmaTuron** [comment]
Hi @EstherIdabor !
Good job, can you add some comments about the performance of your models?

**Comment by EstherIdabor** [comment]
> Hi @EstherIdabor ! Good job, can you add some comments about the performance of your models?

Thank you @GemmaTuron I am on it, because i am very new to machine learning, I am trying hard to understand every step I should take

**Comment by Femme-js** [comment]
Hi @EstherIdabor !

Just to simplify the way in which you can interpret your model, here are some points you can look into:

- Time taken by the model to train
- Area under ROC curve and the range in which it lies, if it's greater than 0.7 your model is fairly good. 
- Checking the number of False Positives, Precision and Recall
- What threshold you chose to covert predicted probabilities into classes and why?

I hope this can help you.

**Comment by EstherIdabor** [comment]
**Evaluation of My Model**
Using the validation set these are the steps I took in evaluating my model:
the validation set was set aside and not used during the training process in order evaluate the performance of my model, by obtaining predictions using the validation and test set, I was able to apply various performance metrics on it and therefore judged how it performed

Using the `predict_proba` function, I got the probability of the prediction of the validation and test molecules being 1(substrate) from whence the probabilty of the prediction not being 1 i.e 0 can also be gotten.

**Evaluation of the model was done using the following metrics;**

- AUC ROC curve
- AUC value
- Confusion Matrix/Contingency Matrix
- Precision
- Recall

**Consequent from the above the following libraries was used;**

- Sckit-Learn
- Matplotlib
- sklearn

**AUROC CURVE**
The ROC curve plots the the true positive(TP) rate versus the false positive rate  at different classification thresholds, TRP is the proportion of positive data points that are correctly predicted as positive. FRP is the proportion of negative data points that are incorrectly  predicted as positives.
The higher the AUC of the model(closer to 1 e.g 0.9) the better the performance of the model in distinguishing between the positive and negative classes.
the lesser the AUC or farther away from 1, the lesser the performance of the model and poor at dicrimination 

values between 0.9-1 is considered excellent, between 0.8-0.9 is considered good, between 0.7-0.8 is considered fair and between 0.6-0.5 is considered poor performance.


****

**Comment by EstherIdabor** [comment]
> Hi @EstherIdabor !
> 
> Just to simplify the way in which you can interpret your model, here are some points you can look into:
> 
> * Time taken by the model to train
> * Area under ROC curve and the range in which it lies, if it's greater than 0.7 your model is fairly good.
> * Checking the number of False Positives, Precision and Recall
> * What threshold you chose to covert predicted probabilities into classes and why?
> 
> I hope this can help you.

Thank you!

**Comment by Zainab-ik** [comment]
@EstherIdabor, The change in AUC is a very significant increase, did that increase occur in the validation set or test set?

**Comment by EstherIdabor** [comment]
> @EstherIdabor, The change in AUC is a very significant increase, did that increase occur in the validation set or test set?

It occurred in the validation set, though I am trying to reason out why it didn't occur like that in the test set

**Comment by EstherIdabor** [comment]
### **CONCLUSION**
**Why is this prediction important?**
Prediction of CYP3A4  substrate is becoming an integral part of  drug research, as its knowledge is important in drug-drug interaction, and drug dosing. Administration of cyp3a4 substrate drug with a cyp3a4 inhibitor drug decreases an individual dosage requirement of a cyp3a4 drug. 
Missing this step in drug research can lead to unintended consequence, this is because administering drug without the knowledge of biotransformation profile can result in the serum level these drugs reaching a toxic state, the toxicity can manifest itself with serious medical consequence. and cause withdrawal from the market.

That said, In enhancing the performance of my model, I took into consideration the training time, as it is necessary to ensure the model is well trained or fitted to make good predictions, 
I also recognize that different learning algorithm can affect the performance as well, depending on the dataset.
Altering the threshold can also help minimize false predictions. 
They are a list of exhaustive metrics but I consider the above best suited to judge the performance of my model.

**Comment by Zainab-ik** [comment]
> ### **CONCLUSION**
> **Why is this prediction important?** Prediction of CYP3A4 substrate is becoming an integral part of drug research, because its knowledge is important in preventing drug-drug interaction, and drug dosing. Administration of cyp3a4 substrate drug with a cyp3a4 inhibitor drug decreases an individual dosage requirement of a cyp3a4 drug. Missing this step in drug research can lead to unintended consequence, this is because administering drug without the knowledge of biotransformation profile can result in the serum level of these drugs reaching a toxic state, the toxicity can manifest itself with serious medical consequence. and can cause withdrawal from the market.
> 
> That said, In enhancing the performance of my model, I took into consideration the training time, as it is necessary to ensure the model is well trained or fitted to make good predictions, I also recognize that different learning algorithm can affect the performance as well, depending on the dataset. Altering the threshold can also help minimize false predictions. They are a list of exhaustive metrics but I consider the above best suited to judge the performance of my model.

Beautiful conclusion @EstherIdabor I'd say. I'd like to add that CYP3A4 is an enzyme involved in drug metabolism but not all classes of drugs. In as much as its knowledge is important to prevent dangerous DDI (drug-drug interaction) and even DFI (drug-food interaction). It's also important to note that it can enhance DDI thereby increasing the therapeutics effects of drugs.

You did a good job.

**Comment by GemmaTuron** [comment]
Hi @EstherIdabor !

thanks for the good job and for re-thinking and trying to improve the performance. 
I'll close this as completed and let you focus on your final application!
